src/transformers/models/deepseek_v3/modular_deepseek_v3.py

```diff
@@ -153,10 +181,114 @@ def moe(self, hidden_states: torch.Tensor, topk_indices: torch.Tensor, topk_weig
         return final_hidden_states.type(hidden_states.dtype)
 
 
+def _remove_inference_params(module: nn.Module):
+    assert isinstance(module, DeepseekV3Attention)  # Sanity check
+    module.inference_v_proj = None
+    module.inference_q_decode = None
+
+
+def _full_backward_hook(module: nn.Module, grad_input, grad_output):
+    _remove_inference_params(module)
+
+
+def _load_state_dict_post_hook(module: nn.Module, incompatible_keys):
+    _remove_inference_params(module)
+
+
+class AttentionParametersFingerprint:
+    def __init__(
+        self,
+        q_b_proj: torch.Tensor,
+        kv_b_proj: torch.Tensor,
+        fingerprint_size: int = 256,
+    ):
+        self.q_b_shape = q_b_proj.shape
+        self.kv_b_shape = kv_b_proj.shape
+        q_b_proj = q_b_proj.flatten()
+        kv_b_proj = kv_b_proj.flatten()
+        q_b_sz = q_b_proj.shape[0]
+        kv_b_sz = kv_b_proj.shape[0]
+        assert fingerprint_size >= 2
+        fp_q_sz = min(fingerprint_size // 2, q_b_sz)
+        fp_kv_sz = min(fingerprint_size - fp_q_sz, kv_b_sz)
+        self.device = q_b_proj.device
+        self.fp_q_pos = torch.randint(
+            low=0,
+            high=q_b_sz,
+            size=(fp_q_sz,),
+            dtype=torch.int64,
+            device=self.device,
+        )
+        self.fp_kv_pos = torch.randint(
+            low=0,
+            high=kv_b_sz,
+            size=(fp_kv_sz,),
+            dtype=torch.int64,
+            device=self.device,
+        )
+        self.fingerprint = self._extract_fingerprint(q_b_proj, kv_b_proj).clone()
+
+    def match(self, q_b_proj: torch.Tensor, kv_b_proj: torch.Tensor) -> bool:
+        if q_b_proj.shape != self.q_b_shape or kv_b_proj.shape != self.kv_b_shape:
+            return False
+        if q_b_proj.device != self.device or kv_b_proj.device != self.device:
+            return False
+        args_fingerprint = self._extract_fingerprint(
+            q_b_proj.flatten(), kv_b_proj.flatten()
+        )
+        return self.fingerprint.eq(args_fingerprint).all().item()
+
+    def _extract_fingerprint(
+        self, q_b_proj: torch.Tensor, kv_b_proj: torch.Tensor
+    ) -> torch.Tensor:
+        assert q_b_proj.ndim == kv_b_proj.ndim == 1
+        assert q_b_proj.device == kv_b_proj.device == self.device
+        return torch.cat((q_b_proj[self.fp_q_pos], kv_b_proj[self.fp_kv_pos]))
```

Comment:
I don't know if this is related or not?

Reply:
Hello @bzantium , @ArthurZucker . I am very happy to delay my contribution and get the current state in. Shall we do that? I'd appreciate your help on the 2nd PR, especially when it comes to testing things on GPUs. As you can see, I wrote some basic comparison tests, but that is all I ran so far.

Reply:
@ArthurZucker The idea here is that there are some derived set of weights for the "inference" implementation. I need to make sure that these are recomputed when the primary weights change, but only then. I thought for a bit how to do that, and this was the best I came up with: Using some of the hooks (mainly for backward load_weights), plus this fingerprint check.

But here, of course, I'd appreciate hints how one could do this better. This is essentially how to detect when model weights have been changed or not.

Reply:
And the changes in the Llama code are mainly about handling the case when queries and keys have a different dimension 1, which happens in MLA, but also in grouped attention. In this case, you don't want to blow up the dimension of keys, but instead for example the `eager_attention_forward` should use batched `matmul`.

This could indeed be split out as well, but I think it is relevant for MLA in the inference case, because the goal there is to safe as much memory as possible.